src/mds/MDSDaemon.cc

```diff
@@ -781,56 +774,82 @@ void MDSDaemon::handle_mds_map(const cref_t<MMDSMap> &m)
 	    << " not writeable with daemon features " << mdsmap_compat
 	    << ", killing myself" << dendl;
     suicide();
-    goto out;
-  }
-
-  // mark down any failed peers
-  for (const auto &p : oldmap->get_mds_info()) {
-    if (mdsmap->get_mds_info().count(p.first) == 0) {
-      dout(10) << " peer mds gid " << p.first << " removed from map" << dendl;
-      messenger->mark_down_addrs(p.second.addrs);
-    }
+    return;
   }
 
-  // see who i am
-  dout(10) << "my gid is " << monc->get_global_id() << dendl;
+  // Calculate my effective rank (either my owned rank or the rank I'm following if STATE_STANDBY_REPLAY
+  const auto addrs = messenger->get_myaddrs();
+  const auto myid = monc->get_global_id();
+  const auto mygid = mds_gid_t(myid);
+  const auto whoami = mdsmap->get_rank_gid(mygid);
+  const auto old_state = oldmap->get_state_gid(mygid);
+  const auto new_state = mdsmap->get_state_gid(mygid);
+  const auto incarnation = mdsmap->get_inc_gid(mygid);
+  dout(10) << "my gid is " << myid << dendl;
   dout(10) << "map says I am mds." << whoami << "." << incarnation
 	   << " state " << ceph_mds_state_name(new_state) << dendl;
+  dout(10) << "msgr says I am " << addrs << dendl;
+
+  // If we're removed from the MDSMap, stop all processing.
+  using DS = MDSMap::DaemonState;
+  if (new_state == DS::STATE_NULL && old_state != DS::STATE_NULL) {
+    const auto& oldinfo = oldmap->get_info_gid(mygid);
+    const auto existing = mdsmap->find_mds_gid_by_name(name);
+    if (g_conf()->mds_enforce_unique_name && existing != MDS_GID_NONE) {
+      const auto& info = mdsmap->get_info_gid(existing);
+      dout(1) << "Map replaced me " << oldinfo
+              << " with another MDS of the same name: " << info
+              << "; quitting!" << dendl;
+
+      /* Monitors should never go backwards! */
+      ceph_assert(info.global_id > myid);
```

Comment:
not sure this is true; the global id is unique but not necessarily monotonic, since the ids can be assigned by any mon in the quorum.

Reply:
For posterity: this code is removed in a subsequent commit.